Gracefully cope with chain forks and the inherent distributed nature of the universe.

KtorZ opened the pull request:
cc @abailly-iohk  I didn't really got time to get to the bottom of the story regarding the necessary modifications in the direct-chain component as I ended up shaving some yak :cow2:. BUT, I did prepare the ground by isolating the relevant piece of code and setting up a little "framework" for testing the chain sync handler with a first test to illustrate it with the `RollForward`. 

Now remains for tomorrow pairing to test that the behavior of the chain-sync handler regarding rollbacks, which should force us to implement the missing pieces. 

---

- :round_pushpin: **Introduce ChainEvent sumtype to also callback on rollback**
  
- :round_pushpin: **Fix ChainEvent dependencies in test**
  
- :round_pushpin: **Write simple property test to show rollback aren't yet handled by the head logic.**
  
- :round_pushpin: **Make the test pass**
    Doesn't do much yet, simply notify clients that something happned.

- :round_pushpin: **Re-generate golden JSON-schemas and fix api specifications.**
  
- :round_pushpin: **Start grooming rollback tests in the behavior specs.**
  
- :round_pushpin: **Make Behavior test compile**
  
- :round_pushpin: **[wip] Modify test chain to drop rolled back txs**
  
- :round_pushpin: **Make HeadState a recursive structure pointing to the previous state from last on-chain event.**
  
- :round_pushpin: **Make test compile but rolling back still failing**
  Had to adapt interface of mock chain to be able to tweak history.

- :round_pushpin: **Implement rollback in the head-logic.**
  
- :round_pushpin: **replay rolled back transactions in our mock-chain.**
    This is one sensible default behavior as it is also the most probable on-chain. However, we will also want to test for other scenarios where for example, some of the events aren't fully replayed.

- :round_pushpin: **Refactor chain-sync client to extract the block handler logic**
    To make it more testable.

- :round_pushpin: **Document previous state and new on-chain events structure in the logs spec.**
  
- :round_pushpin: **re-generate golden test for 'Event'**
    The introduction of the Observation|Rollback sum-type instead of 'OnChainTx' causes this.

- :round_pushpin: **Write simple test to excercise the ChainSyncHandler on roll forward**
  
- :round_pushpin: **Fix integration tests following change in Direct chain**
  
- :round_pushpin: **Small refactor in the chain sync client.**
  
- :round_pushpin: **Start fleshing out rollback test case for ChainSyncHandler.**
  
- :round_pushpin: **Implement basic property with a single block/tx**
  
- :round_pushpin: **Make rollback test fail.**
  
- :round_pushpin: **Start keeping track of points at which state transitions are happening.**
  
- :round_pushpin: **Compare ChainPoints between observed state and rollback point**
  This introduces an orphan for ChainPoint but yolo

- :round_pushpin: **[wip] Pass ChainPoint to observe tx**
  
- :round_pushpin: **Pass block point when processing tx during rollforward.**
  
- :round_pushpin: **Store ChainPoint alongside OnChainHeadState**
  This reverts commit 088e2da7f01ed94233bc53f91d209dd5caff0869.

- :round_pushpin: **Fix rollback point generation.**
  
- :round_pushpin: **Generate longer chains for testing rollbacks.**
  
- :round_pushpin: **Generate longer chains for testing rollbacks.**
  
- :round_pushpin: **Fix commit generation in rollback specs.**
    This is a bit annoying but commits needs to be made 'from the point of view' of each participant, thus requiring us to generate multiple OnChainHeadState (one for each participant) to record commit transactions. This is somewhat 'logical' when one think about it (since we're emulating the chain, with which multiple actors interact) -- but still quite involved.

- :round_pushpin: **Store on-chain head states as recursive structures, pointing back to previous states.**
  
- :round_pushpin: **Slightly rewrite rollback specs to improve readability and error output.**
  
- :round_pushpin: **Define 'Eq' instance for on-chain head state(s) and state-machine.**
  
- :round_pushpin: **Clean-up rollback implementation and adds some explanations here and there.**

ch1bo committed:
Introduce ChainEvent sumtype to also callback on rollback

abailly committed:
Fix ChainEvent dependencies in test

KtorZ committed:
Write simple property test to show rollback aren't yet handled by the head logic.

KtorZ committed:
Make the test pass

  Doesn't do much yet, simply notify clients that something happned.

KtorZ committed:
Re-generate golden JSON-schemas and fix api specifications.

KtorZ committed:
Start grooming rollback tests in the behavior specs.

abailly committed:
Make Behavior test compile

abailly committed:
[wip] Modify test chain to drop rolled back txs

KtorZ committed:
Make HeadState a recursive structure pointing to the previous state from last on-chain event.

abailly committed:
Make test compile but rolling back still failing

Had to adapt interface of mock chain to be able to tweak history.

KtorZ committed:
Implement rollback in the head-logic.

KtorZ committed:
replay rolled back transactions in our mock-chain.

  This is one sensible default behavior as it is also the most probable on-chain. However, we will also want to test for other scenarios where for example, some of the events aren't fully replayed.

KtorZ committed:
Refactor chain-sync client to extract the block handler logic

  To make it more testable.

KtorZ committed:
Document previous state and new on-chain events structure in the logs spec.

KtorZ committed:
re-generate golden test for 'Event'

  The introduction of the Observation|Rollback sum-type instead of 'OnChainTx' causes this.

KtorZ committed:
Write simple test to excercise the ChainSyncHandler on roll forward

abailly committed:
Fix integration tests following change in Direct chain

KtorZ committed:
Small refactor in the chain sync client.

KtorZ committed:
Start fleshing out rollback test case for ChainSyncHandler.

abailly committed:
Implement basic property with a single block/tx

KtorZ committed:
Make rollback test fail.

KtorZ committed:
Start keeping track of points at which state transitions are happening.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -285,15 +283,64 @@ client tracer point queue headState callback nodeToClientV =
+  -- | A chain-sync handler to use in a local-chain-sync client.
+  m (ChainSyncHandler m)
+newChainSyncHandler tracer callback headState = do
+  pure $
```
Why is this function monadic? AFAIS this could be a pure (smart) constructor of `ChainSyncHandler`

KtorZ replied:
No :no_mouth: , but the monadic nature was indeed done in anticipation.

KtorZ replied:
OKAY! It's not needed in the end.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -285,15 +283,64 @@ client tracer point queue headState callback nodeToClientV =
+    -- TODO: need to compute how much to rollback
+    callback (Rollback 0)
+
+  withNextTx :: [OnChainTx Tx] -> ValidatedTx Era -> STM m [OnChainTx Tx]
```
This is a better name :+1:

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -181,75 +200,96 @@ update Environment{party, signingKey, otherParties} ledger st ev = case (st, ev)
+        { parameters
+        , pendingCommits = remainingParties
+        , committed = newCommitted
+        , previousState
```
All the `previousState@` binds can be avoided by defining `previousState = st`, which we already have in scope.

KtorZ replied:
Yes, but we purposely did that because the `st` is a 2-letter identifier which is 500 lines above; so when looking at the branches in isolation, it really comes out of context and, having the rebind is cost-free and immediately clarify what this thing is.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -181,75 +200,96 @@ update Environment{party, signingKey, otherParties} ledger st ev = case (st, ev)
+                        { seenTxs = newSeenTxs
+                        , seenUTxO = utxo'
+                        }
+                  , previousState
```
So the `previousState` is only updated on transitions? This would not be clear to me from the name and I would highly appreciate a haddock string in the definition of `HeadState` explaining what `previousState` is.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -393,14 +454,35 @@ newSn Environment{party} parameters CoordinatedHeadState{confirmedSnapshot, seen
   st -> (st, effects)
+
+rollback :: HasCallStack => Word -> HeadState tx -> HeadState tx
+rollback n
```
So this seems to rely on the fact that `previousState` is updated on every chain transaction seen in sequence. This means also, that the number present here needs to correspond to the number of state transitions.. but how would the chain layer know what how many transitioning events it emited to the head logic?

Just writing down my thoughts as you are probably working on this right now and I'm interested in seeing how you guys solve this :)

ch1bo replied:
This function needs some documentation. It should explain or point to the other function which comes up with the `Word` of steps to rollback

abailly committed:
Compare ChainPoints between observed state and rollback point

This introduces an orphan for ChainPoint but yolo

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -393,12 +428,13 @@ waitUntil nodes expected =
 -- | A thin layer around 'HydraNode' to be able to 'waitFor'.
 data TestHydraNode tx m = TestHydraNode
   { send :: ClientInput tx -> m ()
+  , chainEvent :: ChainEvent tx -> m ()
```
Seems like the implementation ignores `Observation` chain events (makes sense), so I would suggest we make it obvious that we only can issue rollbacks with this function by using the signature:

```hs
, chainRollback :: Word -> m ()
```

or whatever the payload of `Rollback` be by the time you reading this.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -144,6 +164,49 @@ spec = parallel $ do
+            run $ mapM_ (onRollForward handler) blks
+            st' <- run $ readTVarIO headState
+            monitor $ counterexample $ "On-chain head state: " <> show st'
+            run $ onRollBackward handler rollbackPoint
```
I fail to see how this test is checking that "rollback rewind the chain state"? Maybe the naming is wrong and what we test is that the `callback` is calling with `Rollback` when the chain sync does a `onRollBackward`?

ch1bo reviewed (commented):
I'm not yet convinced by the linked list, but interested to see how this will work :)

abailly committed:
[wip] Pass ChainPoint to observe tx

KtorZ committed:
Pass block point when processing tx during rollforward.

abailly committed:
Store ChainPoint alongside OnChainHeadState

This reverts commit 088e2da7f01ed94233bc53f91d209dd5caff0869.

KtorZ committed:
Fix rollback point generation.

KtorZ committed:
Generate longer chains for testing rollbacks.

KtorZ committed:
Generate longer chains for testing rollbacks.

KtorZ committed:
Fix commit generation in rollback specs.

  This is a bit annoying but commits needs to be made 'from the point of view' of each participant, thus requiring us to generate multiple OnChainHeadState (one for each participant) to record commit transactions. This is somewhat 'logical' when one think about it (since we're emulating the chain, with which multiple actors interact) -- but still quite involved.

KtorZ committed:
Store on-chain head states as recursive structures, pointing back to previous states.

KtorZ committed:
Slightly rewrite rollback specs to improve readability and error output.

KtorZ committed:
Define 'Eq' instance for on-chain head state(s) and state-machine.

KtorZ committed:
Clean-up rollback implementation and adds some explanations here and there.

KtorZ commented:
Fix #296
Fix #306 
Fix #307

KtorZ commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -285,15 +294,100 @@ client tracer point queue headState callback nodeToClientV =
+      AtPoint at previous ->
+        if at <= pt
+          then (st, n)
+          else backward (succ n) previous
```
My brain is fried, but we should be careful about this function and likely revisit it or carefully evaluate the consequences of it. While simple in appearance, this might subtly create a data-inconsistency between the chain-sync server and our client. 

Indeed, the server asks us to roll back to point X, and we actually rollback to a point **prior to** X. Thus, when the server start again streaming block, it'll do so from X+1, and so on. Thus, we end up with a (at least mental) gap between where we're at and where the server is at. 

I don't think this matter in practice, because we know that by construction, there is _no observable transaction_ between where we're at (i.e. the last known state) and where the server will continue streaming block from (i.e. the point of rollback). This would however definitely look weird if we started logging all blocks that we see; the gap would be visible. 

So, probably nothing, but maybe worth a comment and definitely worth revisiting with less caffeine and more sleep in the body.

ch1bo replied:
The whole concept also seems to rely on the fact take both layers are consistent in how many chain states they have in their respective lists. While this (seemingly) does work, it is quite fragile and I would propose to have instead the more complex callback, but single state ADR18 (in a follow-up iteration of course)

KtorZ replied:
On the one hand, there's really nothing forcing us to represent states on both side in the same manner, we just happen to use the same approach because it's the most obvious one and simple one.

On the other hand, I do agree with you and it makes me uneasy to have pretty much the same logic on both ends.

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/Point.hs:
```diff
@@ -14,3 +17,18 @@ toConsensusPointHF ::
+  ChainPoint
+fromConsensusPointHF Consensus.GenesisPoint = ChainPointAtGenesis
+fromConsensusPointHF (Consensus.BlockPoint slot (Consensus.OneEraHash h)) =
+  ChainPoint slot (fromJust $ deserialiseFromRawBytes (proxyToAsType Proxy) (fromShort h))
```
Why is `fromJust` okay here? Maybe we could avoid partial functions?

KtorZ replied:
The reason behind this one is because cardano-api does not expose the necessary constructors to construct hashes. So, we could try to make a PR upstream to either expose the constructors or give smart constructors. In the meantime though, there isn't much way around it.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -59,12 +65,42 @@ deriving instance IsTx tx => Show (Effect tx)
 
+-- | A recursive data-structure which represent the application state as a
+-- state-machine. The 'previousRecoverableState' records the state of the
+-- application before the latest 'OnChainEffect' that has been observed.
```
`OnChainEvent` .. events are observed

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -144,13 +178,157 @@ spec = parallel $ do
+    initTx <-
+      lift $
+        initialize (ctxHeadParameters ctx) (ctxVerificationKeys ctx)
+          <$> genTxIn
```
Only `genTxIn` is monadic and needs `lift`. Maybe easier to read if we do `txIn <- lift $ genTxIn` before the pure `initialize`.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -398,3 +586,21 @@ instance Enum Transition where
+
+showRollbackInfo :: (Word, Point Block) -> String
+showRollbackInfo (rollbackDepth, rollbackPoint) =
+  toString $
```
Should we use `Text` or `String` in `showXXX` functions? It just occurred to me because this and the function below differ.

KtorZ replied:
And... Sadly, it's not really our fault. Relude, cardano-api and few other libraries we use work with Text. But Quickcheck and hspec are stuck with String, so there's a bit of back and forth in our codebase.

Note that un #312, I've tried to make the status quo a bit better by making a few of such functions polymorphic in the return type, so String or Text is picked depending on the context.

ch1bo reviewed (changes requested):
Interesting idea with the linked lists. Appears a bit fragile to me, but if it works I'm fine with it for the time being. Before merge, some additional documentation on the two `rollback` functions / tying them together and stating the dependency should be added.

KtorZ committed:
Add missing 'SlotNo' file from 'cardano-api'

KtorZ committed:
Remove redundant import.

abailly committed:
Fix various compilation and test issues

abailly committed:
Provide comments on key rollback functions

ch1bo committed:
Update version numbers

abailly committed:
Update Changelog

ch1bo committed:
Fix comment

ch1bo committed:
Rename newChainSyncHandler -> chainSyncHandler

And make it non-monadic

ch1bo committed:
Apply hlint suggestion for a newtype

ch1bo committed:
Tiny refactor on a lift

ch1bo committed:
Another hlint suggestion
